packages/core/src/Queue.js

```diff
@@ -1,64 +1,98 @@
-/** @license MIT License (c) copyright 2010-2016 original author or authors */
-/** @author Brian Cavalier */
-/** @author John Hann */
+/** @license MIT License (c) copyright 2010 original author or authors */
 
-// Based on https://github.com/petkaantonov/deque
+// Based on https://github.com/invertase/denque
 
 export default class Queue {
-  constructor (capPow2 = 32) {
-    this._capacity = capPow2
-    this._length = 0
+  constructor () {
     this._head = 0
+    this._tail = 0
+    this._capacityMask = 0x3
+    this._list = new Array(4)
   }
 
   push (x) {
-    const len = this._length
-    this._checkCapacity(len + 1)
+    const tail = this._tail
+    this._list[tail] = x
+    this._tail = (tail + 1) & this._capacityMask
+    if (this._tail === this._head) {
+      this._growArray()
+    }
 
-    const i = (this._head + len) & (this._capacity - 1)
-    this[i] = x
-    this._length = len + 1
+    if (this._head < this._tail) {
+      return this._tail - this._head
+    } else {
+      return this._capacityMask + 1 - (this._head - this._tail)
+    }
   }
 
   shift () {
     const head = this._head
-    const x = this[head]
+    if (head === this._tail) {
+      return undefined
+    }
+
+    const x = this._list[head]
+    this._list[head] = undefined
+    this._head = (head + 1) & this._capacityMask
+    if (head < 2 && this._tail > 10000 && this._tail <= this._list.length >>> 2) {
```

Comment:
I think it would be useful to assign `10000` to a well-named variable or introduce a comment explaining why `10000`. Both would likely be worthwhile. Right now, `10000` is pure magic.

Reply:
I should probably explain my approach a bit more.  I copied the implementation of denque, and then retained only the methods we need (push and shift, and the ones they call) to keep code size down, and then adjusted the code style to match ours.  I wanted to change the actual implementation as little as possible for a couple reasons:

1. modifying the code of algorithms like this require a lot of care, since they rely on subtle things like bitwise ops, and
2. if there are changes to denque, we can incorporate them more easily

That said, I don't mind making very small tweaks, like extracting constants, if we feel it's worthwhile.

Reply:
I feel `10000` is a magic number. My immediate questions are: Why `10000`? Why not `32768` or `16384`?

Thus, I think that an explanation of `10000` is appropriate in the code.

Reply:
To be honest: my explanation is that I'm assuming the denque package has made the decision to use 10000 for a reason.  I've literally copied and pasted this code, only changing formatting :)

If we added denque as a dependency, we'd be in basically the same situation.  The reason I've not added it as a dependency is that we only need about 25% of the code in it, and it's prototype-based, so we'd have no choice but to end up with 100% of it in our build.

Reply:
I think copy & modify is a good approach because we get to _own_ the code. Owning code is important because it gives absolute control. Now, since we at this point are in control, we should definitely document the `10000` magic number; at least as a constant.